runtime/src/main/java/org/corfudb/runtime/collections/StreamSubscription.java

```diff
@@ -127,8 +132,19 @@ boolean enqueueStreamEntry(ILogData logData, long maxBlockTime) throws Interrupt
                     .stream()
                     .map(entry -> CorfuStreamEntry.fromSMREntry(entry, epoch))
                     .collect(Collectors.toList());
+
+            // Deduplicate entries per stream Id
+            Map<Message, CorfuStreamEntry> observedKeys = new LinkedHashMap<>();
+            entryList.forEach(entry -> {
+                if (observedKeys.putIfAbsent(entry.getKey(), entry) != null) {
+                    // If the key is present, remove and insert again to preserve transaction order
+                    observedKeys.remove(entry.getKey());
+                    observedKeys.put(entry.getKey(), entry);
+                }
```

Comment:
Why not just do 

```java

Map<Message, CorfuStreamEntry> observedKeys = new LinkedHashMap<>();
            entryList.forEach(entry -> {
                observedKeys.put(entry.getKey(), entry);
            });

```


Latest keys will overwrite the older keys that's because the list has then order by version (implicit mutation order). 

Also, instead of doing another iteration, you can construct the latest unique keys in the same stream https://github.com/CorfuDB/CorfuDB/pull/3008/files#diff-2de6eaba4e142127e6a6aec1b2c3e0e5c13c0312251c321a457f8858722e0fd1R133

Reply:
the problem is that we want to preserve the order, so the case:
k1, v1
k1, v2
k2, v3
k1, v4 -> with this approach would yield:

k1, v4 -> k2, v3 (which is incorrect).. and I tried the 'replace' function as well but that also preserves the initial insertion order